Fix warnings on clang++

schveiguy opened the pull request:
Currently, the default OSX compiler clang reports a boatload of warnings. The vast majority of these warnings have to do with 2 things:
1. `A && B || C` => clang requires `(A && B) || C` even though operator precedence defines this as the default
2. Some issue with using memcpy to copy the vtable of a class instance (this appears a lot because it's defined in a header file).

Seeing as these are mostly just busywork to fix, and not helpful to understanding the code, I disabled these warnings.

In addition, there are several warnings about using switch with enums that require switch be like D's final switch (all enum members accounted for, cannot use values that don't appear in enum). I disabled this as well.

Finally, I did modify some code to comply with some other warnings, because these seem worth while:
1. clang complains that you need to use braces on the outer if statement to avoid confusion with dangling else:

``` C
if(cond1)
    if(cond2)
        statement;
    else
        statement;
```
1. To avoid issues where someone types `a || 5` where they really meant `a | 5`, constants must come at the front of any logical boolean expressions. There are many places where constants are used, and modifying the code did not harm anything (and actually logically reads more optimized).
2. Fixed a few issues where a variable might be uninitialized.
3. Fixed an issue where comparing a char to an unsigned int will always fail.
4. Fixed unsigned comparison less than 0.

Note, there are 2 warnings about unused expressions on some heavy comma operator abuse that I did not touch. Those statements would have to be completely refactored, and I don't feel it's worth the effort. But I did not want to disable that warning.

To run with the newly disabled warnings, you must override the compiler on OSX to be clang++ (clang has issues with the auto tester, and so is not the default on OSX):

`make -f posix.mak HOST_CC=clang++`

Update: This fixes bug https://issues.dlang.org/show_bug.cgi?id=13932

schveiguy committed:
Set up makefile to ignore certain useless warnings for clang (use make -f posix.mak HOST_CC=clang++)
Fix all other warnings that clang identifies except for comma expression abuse

yebblies commented:
Could you please add a test case for the cppmangle issue?

schveiguy commented:
Not sure what you mean?

yebblies commented on src/cppmangle.c:
```diff
@@ -164,7 +164,8 @@ class CppMangleVisitor : public Visitor
                             dinteger_t val = e->toInteger();
-                            if (val < 0)
+                            // TODO: try and do this without knowing what dinteger_t type is
+                            if ((long)val < 0)
```
I mean, this is a bug, right?  Negative C++ template value parameters are currently being mangled incorrectly.  It would be nice to have a test case to go with this fix.

schveiguy replied:
Ah ok I got it. I admit a lot of ignorance in how to add tests for dmd. I will look into it. I was just fixing the warning complaint :)

WalterBright replied:
cast to `sinteger_t` instead, because `val` might be a long long.

schveiguy replied:
Is there a reason val cannot just be declared sinteger_t?

yebblies replied:
No, in fact it has to be otherwise the negation won't work.

schveiguy replied:
> Negative C++ template value parameters are currently being mangled incorrectly

I tried to figure out how to make this happen, I have no idea :(

yebblies replied:
Something like this?

```d
struct S(int x)
{
    int member;
};

extern(C++)
void func(S!(-1) s);

void main()
{
    func(S!(-1)(0));
}
```
```cpp

template<int x>
struct S
{
    int member;
};

void func(S<-1> s) {}
```

And add it to `runnable/cppa`.  If the mangling doesn't match it won't link successfully.

schveiguy replied:
Thank you, that does show the bug, my attempt was way off :) I will add.

schveiguy replied:
> No, in fact it has to be otherwise the negation won't work.

This doesn't make sense to me BTW. The whole point of this code is to unset the sign bit. This should make the unsigned version the same as signed version.

yebblies replied:
Yeah you're right.  It really should be `sinteger_t` though.

schveiguy replied:
> Yeah you're right. It really should be sinteger_t though.

Yep, I will do. Updates coming soon...

Note: I also found a test `runnable/cppmangle1.d` which seems to be more appropriate place for mangling. I have added test to there and verified old compiler doesn't link and new compiler does.

yebblies replied:
No, they usually go into `cppa`.  It doesn't really matter though.

WalterBright commented on src/parse.c:
```diff
@@ -4440,7 +4440,7 @@ void Parser::checkCstyleTypeSyntax(Loc loc, Type *t, int alt, Identifier *ident)
 Statement *Parser::parseStatement(int flags, const utf8_t** endPtr, Loc *pEndloc)
 {
-    Statement *s;
+    Statement *s = NULL;
```
This warning is a false positive. Initializing it to `NULL` can actually hide a bug, as all paths must initialize it. I suggest initializing it instead to `(Statement *)-1`.

yebblies replied:
`cast(Statement)-1` isn't valid D.

schveiguy replied:
How is it a false positive? If I follow the code correctly, at line 5057, when token is `TOKversion`, and next token is `TOKAssign`, `s` is not initialized.

Is this a bug?

Can we do `(Statement *)(void *)1`, would that translate to valid D?

yebblies replied:
Yikes, yes that's a bug, and I think I introduced it.  Those errors where `version = xxx` and `debug = xxx` should probably `goto Lerror` instead of breaking.  Will that be enough to silence the warning?

schveiguy replied:
> Will that be enough to silence the warning?

Yep. That works.

> cast(Statement)-1 isn't valid D.

Is there a sentinel value I can use to assert before the `return s` statement to avoid having to use clang as an error checking tool, and also avoid jumping into undefined memory?

yebblies replied:
Can't you just return from `Lerror` and `assert(s != NULL)` before the main return?

schveiguy replied:
There are other cases that set `s` to `NULL`.

yebblies replied:
The other cases all also emit errors and could be changed to return/goto Lerror as far as I can see.

schveiguy replied:
I tried the goto, it works, but it now fails the test diag11198:

```
Test failed.  The logged output:
../src/dmd -m64 -Ifail_compilation -verrors=0   -odtest_results/fail_compilation -oftest_results/fail_compilation/diag11198_0.o -c fail_compilation/diag11198.d
fail_compilation/diag11198.d(11): Error: version conditions can only be declared at module scope


==============================
Test failed: 
expected:
----
fail_compilation/diag11198.d(11): Error: version conditions can only be declared at module scope
fail_compilation/diag11198.d(12): Error: debug conditions can only be declared at module scope
----
actual:
----
fail_compilation/diag11198.d(11): Error: version conditions can only be declared at module scope
----


make[1]: *** [test_results/fail_compilation/diag11198.d.out] Error 1
```

Is this what you would expect to happen?

schveiguy replied:
> The other cases all also emit errors and could be changed to return/goto Lerror as far as I can see.

This is getting out of my comfort zone. I'm thinking I'll fix the bug that the warning found, and you can implement your fix when you have time.

schveiguy replied:
> Is this what you would expect to happen?

Update: instead of `goto Lerror`, I just set `s` to `NULL`, and it seems to work, will continue testing.

yebblies replied:
I'm fine with you just setting it to `NULL` at the top, but the goto should work.

yebblies replied:
Oh, you need to delete both the `check(TOKsemicolon);` and the `break;` lines, because `Lerror` skips up until the next semicolon then skips the semicolon.  So if you have already skipped the semicolon it will completely skip the next statement.

schveiguy replied:
> I'm fine with you just setting it to NULL at the top, but the goto should work.

@WalterBright seems to dislike this, but I can do whatever you guys want. Adding the goto gets rid of the warning, and I will test with the update above. Like I said, I don't really feel comfortable modifying the code, but just doing the minimum possible to fix the warning :)

yebblies replied:
That's fine, I'm happy to merge this once the cppmangle thing is done.

schveiguy committed:
More proper implementation for parse error on version and debug.
Fix issue 13932 -- c++ mangling for template negative int parameter
add test case for 13932

schveiguy commented:
Updated as requested.

yebblies commented:
Is the warning in parse.c still there?

yebblies commented:
Auto-merge toggled on

schveiguy commented:
> Is the warning in parse.c still there?

No, the warning is fixed by the gotos. Only warnings left (and I guess it would be good to post them here) are comma operator abuse:

```
clang++ -m64 -c -Wno-deprecated -Wstrict-aliasing -Wno-logical-op-parentheses -Wno-dynamic-class-memaccess -Wno-switch -fno-exceptions -fno-rtti -D__pascal= -DMARS=1 -DTARGET_OSX=1 -DDM_TARGET_CPU_X86=1  -O2 -DDMDV2=1 -Iroot -Itk -Ibackend -I. -MMD -MF cgelem.deps backend/cgelem.c
backend/cgelem.c:1409:37: warning: expression result unused [-Wunused-value]
            ((e12->Eoper == OP64_32 && ((e12 = e12->E1),1)),
              ~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~
backend/cgelem.c:1418:40: warning: expression result unused [-Wunused-value]
            ((e2111->Eoper == OPu32_64 && ((e2111 = e2111->E1),1)),
              ~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

schveiguy commented:
Failed test seemed to be an environment issue, so I deprecated it.

schveiguy commented:
@yebblies My added test for negative int template parameter is now failing win64! Any ideas?

yebblies commented:
The same error exists for the msvc mangler on line 1483

schveiguy committed:
Fix issue 13932 on Windows as well.

yebblies commented:
> `((e12->Eoper == OP64_32 && ((e12 = e12->E1),1))`

Arrgh my eyes are bleeding.

schveiguy commented:
All set for Windows as well, hopefully it passes...

alexrp commented on src/posix.mak:
```diff
@@ -83,6 +83,12 @@ endif
 else
 # Default Warnings
 WARNINGS := -Wno-deprecated -Wstrict-aliasing
+ifeq ($(HOST_CC), clang++)
```
This isn't going to work if a path is given. I think it would be better to identify the compiler via `$HOST_CC --version` or something.

schveiguy replied:
I just followed precedent, see line 75. Feel free to improve if you wish.

schveiguy commented:
grrr.. more issues with win32 and my new test. This looks weird though, looks like it's incorrectly instantiating the template? (symbol name looks like `func13932(S13932<0>)`, I would expect `func13932(S13932<-1>)` Any ideas @yebblies ?

```
OPTLINK (R) for Win32  Release 8.00.15



Copyright (C) Digital Mars 1989-2013  All rights reserved.



http://www.digitalmars.com/ctg/optlink.html



generated\runnable\cppa_0.obj(cppa_0) 



 Error 42: Symbol Undefined ?func13932@@YAXU?$S13932@$0?0@@@Z (void cdecl func13932(S13932<0> ))
```

yebblies commented:
My guess is it's a difference between dmc and msvc mangling.  @IgorStepanov ?

schveiguy commented:
Would it be possible to commit the changes without the test, and then submit the test as a separate PR to keep this going? This test is really not something I was planning on doing with this PR.

yebblies commented:
Can you disable the test for win32 in the tests makefile?  You just need to add it to DISABLED_TESTS when it's windows.

schveiguy commented:
Hm... I need to do it in the file, I don't want to disable the entire test.

schveiguy committed:
Removing test for 13932 on Win32 -- there is some issue with CPP mangling on optlink.

schveiguy commented:
Once this pulls, I create a PR to remove the version guard, so it's not forgotten. I'll also add another bug report.

joakim-noah commented:
Another issue you may want to look at is that `std.zlib` from phobos segfaults in its unit tests on linux/x86 when etc/c/zlib is compiled by clang.  This has been going on for more than a year, never looked into it further, but it presumably will hit you on OS X too.

schveiguy commented:
@joakim-noah thanks, if there's any way I can look at it, I may. From what I understand, we haven't been using clang for a long time because it fails some tests mysteriously.

schveiguy commented:
While this is waiting, I found some interesting info playing around with dmc.

Apparently, dmc expects the identical name mangling for the following 2 functions:

```D
struct S1(int x)
{
   int member;
}

extern(C++) void func(S1!(-1));

struct S2(ulong x)
{
   int member;
}

extern(C++) void func(S2!(ulong.max));
```
Yep, the symbol according to optlink is:

```
?func@@YAXU?$S@$0PPPPPPPPPPPPPPPP@@@@Z (void cdecl func(S<4294967295> ))
```

So that's why it's not linking now. The problem here is, it does link with 2.066.1, so something I did messed it up. There's a possibility this will be a regression...

yebblies commented:
Oh...  So does that mean dmc always mangles integer template value parameters as unsigned?

schveiguy commented:
to be honest, I don't know. What I showed is what optlink spits out when it can't link. I don't know what all the 'P's mean, or anything else. But the fact that it links if I trick it from the D side into doing `ulong.max` instead of -1, seems to indicate there is some truth to that.

I think it's highly unlikely that dmc would be changed at this point. What we probably should do is version the check for negativity out of Win32.

yebblies commented:
I think the 'P's are the way the number is encoded, in some weird base.  If you try `-1` you should see it off by one.

Yeah, I think versioning it out for win32/dmc is the correct way.

schveiguy commented:
> If you try -1 you should see it off by one

Yep, I see that, it's PPPP...PO :)

I will do another pull with the fix.
